### PR TITLE
TF-2628 Add open downloaded file in new tab feature

### PR DESCRIPTION
**Pull request by tddang-linagora**
### **Issue**
#2628 
### **Resolved**

As discussed with @hoangdat , @dab246 & @hieutbui , attachment button for PDF in web version will have 2 functionalities. When user tap on download icon inside the button, `openDownloadedFileWeb` will be triggered, showing the PDF inside new browser tab and user can download the file here themselves. When tapping anywhere else inside the attachment button, the PDF will be downloaded automatically, and the auto opening feature will be handle by the browser itself depends on its own settings.

- Firefox

https://github.com/linagora/tmail-flutter/assets/160106668/3afedf5d-d3ba-44c2-9130-78b3bc074aa5

- Safari

https://github.com/linagora/tmail-flutter/assets/160106668/60350866-777b-4de4-8bdb-5304128a75e5

- Chrome

https://github.com/linagora/tmail-flutter/assets/160106668/02cbdbe0-538b-47dd-80f3-7f4362393a33

### CODE REVIEW

File: core/lib/data/network/download/download_manager.dart [hunk 1]
```diff
@@ -104,6 +105,21 @@ class DownloadManager {
...
+    try {
+      final mime = mimeType ?? lookupMimeType('', headerBytes: bytes);
+      final blob = html.Blob([bytes], mime);
+      final url = html.Url.createObjectUrlFromBlob(blob);
+
+      html.window.open(url, '_blank');
+
+      html.Url.revokeObjectUrl(url);
+    } catch (exception) {
+      log('DownloadManager::openDownloadedFileWeb(): ERROR: $exception');
```
Comment (dab246) [inline core/lib/data/network/download/download_manager.dart:118]: Please use `logError`

**Comment by dab246** [comment]
According to the video, it seems like we only opened the pdf file in a new tab and did not save it to the device? If so, is there a way to both 'save' and open in a 'new tab'?

**Comment by tddang-linagora** [comment]
> According to the video, it seems like we only opened the PDF file in a new tab and did `not save` it to the device? If so, is there a way to both `save` and open in a `new tab`?

Let me have a look at it

File: pubspec.yaml [hunk 2]
```diff
@@ -229,6 +229,8 @@ dependencies:
 
   flutter_secure_storage: 8.0.0
 
+  mime: 1.0.4
```
Comment (hoangdat) [inline pubspec.yaml:232]: we really need to add it here?

Comment (tddang-linagora) [reply pubspec.yaml:232]: Will check. Thank you.

File: model/lib/email/attachment.dart [hunk 3]
```diff
@@ -20,6 +20,7 @@ class Attachment with EquatableMixin {
   final MediaType? type;
   final String? cid;
   final ContentDisposition? disposition;
+  final AttachmentAction attachmentAction;
```
Comment (hoangdat) [inline model/lib/email/attachment.dart:23]: Why we need to care about `UI` in model?

File: lib/features/email/presentation/controller/single_email_controller.dart [hunk 4]
```diff
@@ -89,8 +90,9 @@ import 'package:tmail_ui_user/main/routes/route_navigation.dart';
 import 'package:tmail_ui_user/main/routes/route_utils.dart';
 import 'package:tmail_ui_user/main/utils/app_utils.dart';
 
-class SingleEmailController extends BaseController with AppLoaderMixin {
+enum AttachmentAction { download, view }
```
Comment (dab246) [inline lib/features/email/presentation/controller/single_email_controller.dart:93]: It should be created in a separate file

File: lib/features/email/presentation/controller/single_email_controller.dart [hunk 5]
```diff
@@ -746,16 +763,36 @@ class SingleEmailController extends BaseController with AppLoaderMixin {
...
+
+  void _viewAttachmentForWebSuccessAction(
+    DownloadAttachmentForWebSuccess success,
+  ) {
+    log('SingleEmailController::_viewAttachmentForWebSuccessAction():');
+    _attachmentsToHandleDownloadOrView.remove(success.taskId);
+    final mimeType = success.attachment.type?.mimeType ??
+        lookupMimeType('', headerBytes: success.bytes);
+    if (mimeType != Constant.pdfMimeType) {
+      // As of 23/02/2024, open in new browser tab only support PDF file
```
Comment (dab246) [inline lib/features/email/presentation/controller/single_email_controller.dart:781]: We should not use `comment` in code. If necessary you can write `ADR` to it.

File: lib/features/email/presentation/controller/single_email_controller.dart [hunk 6]
```diff
@@ -746,16 +763,36 @@ class SingleEmailController extends BaseController with AppLoaderMixin {
 
   void _downloadAttachmentForWebSuccessAction(DownloadAttachmentForWebSuccess success) {
     log('SingleEmailController::_downloadAttachmentForWebSuccessAction():');
+    _attachmentsToHandleDownloadOrView.remove(success.taskId);
     mailboxDashBoardController.deleteDownloadTask(success.taskId);
 
     _downloadManager.createAnchorElementDownloadFileWeb(
-        success.bytes,
-        success.attachment.generateFileName());
+        success.bytes, success.attachment.generateFileName());
```
Comment (dab246) [inline lib/features/email/presentation/controller/single_email_controller.dart:770]: Why change it?

File: lib/features/email/presentation/widgets/attachment_item_widget.dart [hunk 7]
```diff
@@ -11,7 +11,8 @@ import 'package:model/email/attachment.dart';
 import 'package:tmail_ui_user/features/email/presentation/extensions/attachment_extension.dart';
 import 'package:tmail_ui_user/features/email/presentation/styles/attachment/attachment_item_widget_style.dart';
 
-typedef OnDownloadAttachmentFileActionClick = void Function(Attachment attachment);
+typedef OnDownloadAttachmentFileActionClick = void Function(
+    Attachment attachment, bool viewOnly);
```
Comment (dab246) [inline lib/features/email/presentation/widgets/attachment_item_widget.dart:15]: Why not use a separate function for `view`

File: model/lib/email/attachment.dart [hunk 8]
```diff
@@ -59,7 +59,8 @@ class Attachment with EquatableMixin {
   }
 
   @override
-  List<Object?> get props => [partId, blobId, size, name, type, cid, disposition];
+  List<Object?> get props =>
+      [partId, blobId, size, name, type, cid, disposition];
```
Comment (dab246) [inline model/lib/email/attachment.dart:63]: why change it?

File: lib/features/email/presentation/controller/single_email_controller.dart [hunk 9]
```diff
@@ -175,6 +179,12 @@ class SingleEmailController extends BaseController with AppLoaderMixin {
     } else if (success is MarkAsStarEmailSuccess) {
       _markAsEmailStarSuccess(success);
     } else if (success is DownloadAttachmentForWebSuccess) {
+      final action = _attachmentsToHandleDownloadOrView[success.taskId];
+      if (action == null) return;
+      if (action == AttachmentAction.view) {
+        _viewAttachmentForWebSuccessAction(success);
+        return;
+      }
```
Comment (dab246) [inline lib/features/email/presentation/controller/single_email_controller.dart:187]: It's quite confusing to me. Is there any other way?

File: lib/features/email/domain/usecases/view_attachment_for_web_interactor.dart [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    AccountId accountId,
+    String baseDownloadUrl,
+    StreamController<Either<Failure, Success>> onReceiveController,
+  ) =>
+      _downloadAttachmentForWebInteractor
+          .execute(taskId, attachment, accountId, baseDownloadUrl,
+              onReceiveController)
+          .map(
+            (result) => result.fold(
+              (l) {
```
Comment (nqhhdev) [inline lib/features/email/domain/usecases/view_attachment_for_web_interactor.dart:30]: IM0: I don't know exactly `l` what is this. Pls rename for this.

File: lib/features/email/presentation/bindings/email_bindings.dart [hunk 11]
```diff
@@ -55,17 +56,17 @@ class EmailBindings extends BaseBindings {
...
+        Get.find<MarkAsEmailReadInteractor>(),
+        Get.find<DownloadAttachmentsInteractor>(),
+        Get.find<DeviceManager>(),
+        Get.find<ExportAttachmentInteractor>(),
+        Get.find<MoveToMailboxInteractor>(),
+        Get.find<MarkAsStarEmailInteractor>(),
+        Get.find<DownloadAttachmentForWebInteractor>(),
+        Get.find<GetAllIdentitiesInteractor>(),
+        Get.find<StoreOpenedEmailInteractor>(),
+        Get.find<ViewAttachmentForWebInteractor>()));
```
Comment (nqhhdev) [inline lib/features/email/presentation/bindings/email_bindings.dart:69]: IMO: Keep old format

**Comment by tddang-linagora** [comment]
Close to create a new PR due to solution's direction changed

**Pull request closed without merging**